### PR TITLE
Introduce UCI protocol handler

### PR DESCRIPTION
**Pull request by amirov-m**

### CODE REVIEW

File: Communication/BotCommunicator.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+namespace Communication
+{
+    // BotCommunicator is the entry point for all commands coming to the bot.
+    // It is responsible for initial instantiation of the provided protocol and
+    // and further communication proxying.
```
Comment (cardinal-evgeniy) [inline Communication/BotCommunicator.cs:5]: and and

File: Communication/Protocols/UCI/Options/SpinOption.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        private int minAllowed;
+        private int maxAllowed;
+        private int value;
+
+        public SpinOption(string name, int minAllowed, int maxAllowed, int value) : base(name, "spin")
+        {
+            if (minAllowed > maxAllowed)
+            {
+                throw new UCIProtocolException($"Minimum allowed value '{minAllowed}' cannot be bigger then the maximum value '{maxAllowed}'.");
```
Comment (cardinal-evgeniy) [inline Communication/Protocols/UCI/Options/SpinOption.cs:13]: than

File: Communication/Protocols/UCI/Options/SpinOption.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            EnsureInAllowedRange(value);
+            this.value = value;
+        }
+
+        private void EnsureInAllowedRange(int value)
+        {
+            if (value < minAllowed || value > maxAllowed)
+            {
+                throw new UCIProtocolException("$Option value '{value}' must be in range [{this.minAllowed}, {this.maxAllowed}].");
```
Comment (SherlockKA) [inline Communication/Protocols/UCI/Options/SpinOption.cs:31]: Is '$' sign placed correctly? Probably outside of  the string

File: Communication/Protocols/UCI/Options/SpinOption.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
+namespace Communication.Protocols.UCI
+{
+    class SpinOption : BaseOption
```
Comment (SherlockKA) [inline Communication/Protocols/UCI/Options/SpinOption.cs:3]: It is completely not clear to me what SpinOption class exists for. For example, the constructor input is ```int minAllowed, int maxAllowed, int value``` and you check if the value is in range. Could you please explain why we require this class in our case?

File: Communication/Protocols/UCI/Options/StringOption.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
+namespace Communication.Protocols.UCI
+{
+    class StringOption : BaseOption
+    {
+        private static string EmptyValue = "<empty>";
```
Comment (SherlockKA) [inline Communication/Protocols/UCI/Options/StringOption.cs:5]: You have started all field names before with lower case, but this with upper case. Why?

File: Communication/Protocols/UCI/Options/SpinOption.cs (hunk 4 above)
Comment (amirov-m) [reply Communication/Protocols/UCI/Options/SpinOption.cs:3]: `SpinOption` is one of the Option types introduced by the UCI Standard
http://wbec-ridderkerk.nl/html/UCIProtocol.html (Search for "* type")

Potential usage examples:
1. A spin option for selecting the target ELO rating that Bot will try to imitate. 
2. Real one: `option name NalimovCache type spin default 1 min 1 max 32`

File: Communication/Protocols/UCI/Options/StringOption.cs (hunk 5 above)
Comment (amirov-m) [reply Communication/Protocols/UCI/Options/StringOption.cs:5]: I was planning to make it `const`. I forgot to add the modifier.

File: Communication/BotCommunicator.cs (hunk 1 above)
Comment (amirov-m) [reply Communication/BotCommunicator.cs]: Fixed.

File: Communication/Protocols/UCI/Options/SpinOption.cs (hunk 3 above)
Comment (amirov-m) [reply Communication/Protocols/UCI/Options/SpinOption.cs]: Fixed.

**Review by SherlockKA — APPROVED** [review]

**Pull request merged**